### ISSUE TITLE
mod_statcache affects file permissions on upload

### ISSUE THREAD

**Issue by dup2** [proftpd/proftpd#675]
### What I Did

We configured mod_statcache to use caching in shared memory.

It seems to affect the permissions of new files created via SFTP/FTPS. We configured an umask of 0077 which is not used when StatCacheEngine is enabled. 

The following fragment is making a difference:

    <IfModule mod_statcache.c>
      StatCacheEngine on
      StatCacheMaxAge 10 0 # cache entries for 10 seconds, do not cache failed lookups
      StatCacheTable /var/run/proftpd/statcache.tab
    </IfModule>

If this is enabled, a file upload will get permission `--wx------`, if this is disabled the file upload will result in a permission `-rw-------` which is expected.

### What I Expected/Wanted

We expected a file permission of 600 for all files due to the setting "Umask 0077" in the server configuration. Without mod_statcache, this works fine - with mod_statcache we get wrong permissions . A quick glance at the mod_statcache code shows that it provides callbacks to various fsio functions, this might affect the outcome of the final chmod/fchmod call or umask setting.

### ProFTPD Version and Configuration

````
Compile-time Settings:
  Version: 1.3.6 (stable)
  Platform: LINUX [Linux 4.4.0-98-generic x86_64]
  Built: Wed Feb 28 2018 14:57:06 CET
  Built With:
    configure  '--prefix=/opt/cargo' '--with-includes=/opt/cargo/include' '--with-libraries=/opt/cargo/lib' '--enable-auth-pam=no' '--disable-ident' '--enable-facl' '--enable-cap' '--enable-nls' '--enable-shared=no' '--enable-static=yes' '--enable-dso=no' '--enable-openssl' '--enable-ctrls' '--enable-sodium' '--enable-redis' '--enable-timeout-idle=300' '--enable-timeout-stalled=1800' '--enable-pcre' '--with-modules=mod_sql:mod_sftp_sql:mod_sftp:mod_sql_postgres:mod_quotatab:mod_quotatab_sql:mod_unique_id:mod_tls:mod_wrap2:mod_wrap2_sql:mod_sql_passwd:mod_ctrls_admin:mod_ban:mod_clamav:mod_exec:mod_deflate:mod_snmp:mod_copy:mod_geoip:mod_ifsession:mod_digest:mod_statcache:mod_tls_redis:mod_auth_otp:mod_diskuse' 'CC=cc' 'CFLAGS=-fstack-protector-all'

  CFLAGS: -g2 -fstack-protector-all -Wall -fno-omit-frame-pointer
  LDFLAGS: -L$(top_srcdir)/lib  -rdynamic -L/opt/cargo/lib -L/usr/lib
  LIBS:  -lpcreposix -lpcre -lssl -lcrypto -lsodium -lcap  -lhiredis  -lGeoIP  -lz  -lssl  -lm -lpq  -lcrypto -lsupp -lcrypt -ldl

  Files:
    Configuration File:
      /opt/cargo/etc/proftpd.conf
    Pid File:
      /var/run/proftpd.pid
    Scoreboard File:
      /opt/cargo/var/proftpd.scoreboard

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    + Controls support
    + curses support
    - Developer support
    - DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    + ncurses support
    + NLS support
    + Redis support
    + Sodium support
    + OpenSSL support
    + PCRE support
    + POSIX ACL support
    + Shadow file support
    + Sendfile support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 300
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 1800
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
````

**Comment by Castaglia** [comment]
Do you only see this behavior for SFTP/SCP uploads, and not FTP/FTPS uploads?

**Comment by dup2** [comment]
It affects SFTP/SCP/FTPS but seems not to affect plain FTP. 

We get `--wx------` for FTPS and `--wx-----T` for SFTP.

Thanks.

**Comment by Castaglia** [comment]
Interesting.  So far, I'm not able to reproduce this locally.  However, I _am_ using a build that contains the fix for [Bug#4345](http://bugs.proftpd.org/show_bug.cgi?id=4345), and I am wondering if that is perhaps related to the behavior you are seeing.

**Comment by Castaglia** [comment]
I was able to finally demonstrate, locally, that indeed the fix for Bug #4345 addresses this issue as well.  Thanks!

**Comment by dup2** [comment]
Thanks!
